Feature swc 13 crear endpoint para enviar partidas disponibles

MatiasScantamburlo opened the pull request:
Listar partidas disponibles **terminado** ¡Espero su revisión! :100:

MatiasScantamburlo committed:
Se crea el endpoint GET /api/lobby para listar las partidas disponibles

- Implementa el endpoint que retorna información sobre partidas no iniciadas
    con lugares disponibles.
- Filtra partidas que no han comenzado y tienen espacio para nuevos jugadores.
- Ordena las partidas por la cantidad de jugadores en orden ascendente.
- Incluye modelos Pydantic para la validación de datos.

También incluye pruebas para verificar el estado del código, la disponibilidad
de partidas y el orden de los jugadores.

MatiasScantamburlo committed:
Se crea el endpoint GET /api/lobby para listar las partidas disponibles

- Implementa el endpoint que retorna información sobre partidas no iniciadas
    con lugares disponibles.
- Filtra partidas que no han comenzado y tienen espacio para nuevos jugadores.
- Ordena las partidas por la cantidad de jugadores en orden ascendente.
- Incluye modelos Pydantic para la validación de datos.

También incluye pruebas para verificar el estado del código, la disponibilidad
de partidas y el orden de los jugadores.

MatiasScantamburlo committed:
feat: Implementa gestión de conexiones WebSocket por lobby

- Añadida la clase ManejadorConexionesLobby para manejar conexiones y
    desconexiones de jugadores en lobbies.
- Implementada la conexión y desconexión de jugadores a través del endpoint
    WebSocket /ws/lobby/{lobby_id}.
- Agregada funcionalidad de comunicación broadcast para notificar a todos los
    jugadores en el mismo lobby cuando un nuevo jugador se une o cuando un
    jugador abandona.

 Cambios a ser confirmados:
	nuevos archivos: actualizar_lobby.py

MatiasScantamburlo committed:
feat: Implementa gestión de conexiones WebSocket por lobby

- Añadida la clase ManejadorConexionesLobby para manejar conexiones y
    desconexiones de jugadores en lobbies.
- Implementada la conexión y desconexión de jugadores a través del endpoint
    WebSocket /ws/lobby/{lobby_id}.
- Agregada funcionalidad de comunicación broadcast para notificar a todos los
    jugadores en el mismo lobby cuando un nuevo jugador se une o cuando un
    jugador abandona.

 Cambios a ser confirmados:
	nuevos archivos: actualizar_lobby.py

MatiasScantamburlo committed:
Revert "Se crea el endpoint GET /api/lobby para listar las partidas disponibles"

This reverts commit 726798060510404fc485f02a0ec0d582fd4569c7.

MatiasScantamburlo committed:
Revert "Se crea el endpoint GET /api/lobby para listar las partidas disponibles"

This reverts commit 726798060510404fc485f02a0ec0d582fd4569c7.

MatiasScantamburlo committed:
feat: Se crea el endpoint GET /api/lobby para listar las partidas disponibles

- Implementa el endpoint que retorna información sobre partidas no iniciadas
    con lugares disponibles.
- Filtra partidas que no han comenzado y tienen espacio para nuevos jugadores.
- Ordena las partidas por la cantidad de jugadores en orden ascendente.
- Incluye modelos Pydantic para la validación de datos.

test: También incluye pruebas para verificar el estado del código, la disponibilidad
de partidas y el orden de los jugadores.

 Cambios a ser confirmados:
	nuevos archivos: listar_partidas_disponibles.py
	nuevos archivos: test_listar_partidas_disponibles.py

MatiasScantamburlo committed:
feat: Se crea el endpoint GET /api/lobby para listar las partidas disponibles

- Implementa el endpoint que retorna información sobre partidas no iniciadas
    con lugares disponibles.
- Filtra partidas que no han comenzado y tienen espacio para nuevos jugadores.
- Ordena las partidas por la cantidad de jugadores en orden ascendente.
- Incluye modelos Pydantic para la validación de datos.

test: También incluye pruebas para verificar el estado del código, la disponibilidad
de partidas y el orden de los jugadores.

 Cambios a ser confirmados:
	nuevos archivos: listar_partidas_disponibles.py
	nuevos archivos: test_listar_partidas_disponibles.py

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-13

- Se integró la configuración de base de datos desde la rama 'develop'.
- Con el objetivo de remplazar las variables locales utilizadas para almacenar
    la información de las partidas por la nueva configuración de base de datos.
- Preparado el entorno para soportar las nuevas funcionalidades relacionadas
    con la base de datos.

MatiasScantamburlo committed:
Merge branch 'develop' into feature_SWC-13_Crear_endpoint_para_enviar_partidas_disponibles

MatiasScantamburlo committed:
Modulariza funciones y organiza estructura del proyecto

- Se modularizaron las funciones creadas, moviendo las características al
    archivo main y sus respectivos tests a la carpeta designada.
- Se implementó la lógica para consultar e impactar en la base de datos.
- Se agregó el método "get_available" a la clase GameRepository.

 Cambios a ser confirmados:
	borrados:        listar_partidas_disponibles.py
	modificados:     main.py
	modificados:     model/game.py
	modificados:     repositories/game.py
	borrados:        test_listar_partidas_disponibles.py
	modificados:     tests/test_sanity.py

MatiasScantamburlo committed:
Modulariza funciones y organiza estructura del proyecto

- Se modularizaron las funciones creadas, moviendo las características al
    archivo main y sus respectivos tests a la carpeta designada.
- Se implementó la lógica para consultar e impactar en la base de datos.
- Se agregó el método "get_available" a la clase GameRepository.

 Cambios a ser confirmados:
	borrados:        listar_partidas_disponibles.py
	modificados:     main.py
	modificados:     model/game.py
	modificados:     repositories/game.py
	borrados:        test_listar_partidas_disponibles.py
	modificados:     tests/test_sanity.py

MatiasScantamburlo committed:
bug: Quitamos campos host y ongame

- Despues vemos si son necesarios

 Cambios a ser confirmados:
	modificados:     model/player.py

MatiasScantamburlo committed:
bug: Quitamos campos host y ongame

- Despues vemos si son necesarios

 Cambios a ser confirmados:
	modificados:     model/player.py

MatiasScantamburlo committed:
Resuelto conflicto

KaiserOfNone commented on tests/test_sanity.py:
```diff
@@ -6,8 +6,23 @@
+def test_estado_codigo():
+    response = client.get("/api/lobby")
+    assert response.status_code == 200
 
-def test_borrame():
-    response = client.get("/api/borrame")
-    asserts.assert_equal(response.status_code, 200)
-    asserts.assert_equal(response.json(), {"games": []})
+def test_partidas_disponibles():
+    response = client.get("/api/lobby")
+    date = response.json()
+    for party in date:
+        assert party['started'] == False
+        assert len(party['players']) < party['max_players']
+
+def test_orden_cantidad_jugadores():
+    response = client.get("/api/lobby")
+    data = response.json()
+
+    players_count = [len(lobby['players']) for lobby in data]
+    
+    assert players_count == sorted(players_count)
+    
```
Esto no va en los tests de sanidad, movelos a `test_lobbies.py` o algo asi.

KaiserOfNone commented on repositories/game.py:
```diff
@@ -18,3 +18,7 @@ def get(self, id: int) -> Optional[Game]:
     def get_many(self, count: int) -> List[Game]:
         return self.db.query(Game).limit(count).all()
+
+    def get_available(self, count: int) -> List[Game]:
+        games = self.db.query(Game).filter(Game.started == False).all()
+        return [game for game in games if len(game.players) < game.max_players][:count]
```
Saltate el paso de descargar todas las partidas para filtrar las que si queres ver.
(tenes que importar `sqlalchemy.func`)
```suggestion
    def get_available(self, count: int = None) -> List[Game]:
        q = (
            self.db.query(Game)
            .outerjoin(Game.players)
            .group_by(Game.id)
            .having(func.count(Player.id) < Game.max_players)
            .filter(Game.started == False)
        )
        if count is not None:
            q = q.limit(count)
        return q.all()
```
Es mas codigo pero es mucho mejor, en lugar de descargar todo filtras en la db

MatiasScantamburlo replied:
> Saltate el paso de descargar todas las partidas para filtrar las que si queres ver. (tenes que importar `sqlalchemy.func`)
> 
> Es mas codigo pero es mucho mejor, en lugar de descargar todo filtras en la db

¡Muy buen uso de having! supongo que puedo hacer un `.order_by(func.count(Player.id))` para ordenarlos por cantidad de jugadores en lobby

KaiserOfNone replied:
Sip, exactamente

KaiserOfNone commented on repositories/game.py:
```diff
@@ -18,3 +18,7 @@ def get(self, id: int) -> Optional[Game]:
     def get_many(self, count: int) -> List[Game]:
         return self.db.query(Game).limit(count).all()
+
+    def get_available(self, count: int) -> List[Game]:
```
Esta funcion no esta testeada, agrega un test en `repositories/test_game.py`, aca tenes una test que checkea el comportamiento que queres creo

```python
    def test_get_available_games(self):
        repo = self.repo()
        games = [
            Game(name="game1", started=False, players=[Player(name=f"{n}") for n in range(2)], max_players=4),
            Game(name="game2", started=False, players=[Player(name=f"{n}") for n in range(4)], max_players=4),
            Game(name="game3", started=True, players=[Player(name=f"{n}") for n in range(2)], max_players=4),
            Game(name="game4", started=False, players=[Player(name=f"{n}") for n in range(1)], max_players=2)
        ]
        for game in games:
            game.set_defaults()
            repo.save(game)

        available_games = repo.get_available(4)
        asserts.assert_in(games[0], available_games)
        asserts.assert_in(games[3], available_games)
        asserts.assert_not_in(games[1], available_games)
        asserts.assert_not_in(games[2], available_games)
```

MatiasScantamburlo replied:
¡Me re sirve! No sé si puedo usar self en este contexto pero ahí pruebo. De todas forma me sirve. Muchas gracias

KaiserOfNone replied:
Las pruebas en repositorios usan la clase de unittest

https://docs.python.org/3/library/unittest.html#unittest.TestCase

En este caso es organizativo pero podes hacer un `setUp` para levantar y bajar recursos, muy util.

KaiserOfNone commented on model/game.py:
```diff
@@ -68,6 +68,8 @@ def add_player(self, player):
             raise GameFull
         self.players.append(player)
 
+    def count_players(self) -> int:
+        return len(self.players)
```
Esto no hace falta. Mirar `len(game.players)` no es pecado.

KaiserOfNone commented on tests/test_sanity.py:
```diff
@@ -6,8 +6,23 @@
 
 client = TestClient(app)
 
+# ------- LISTAR PARTIDAS DISPONIBLES ------- #
```
No uses "banners" de comentarios, no organizan el codigo. Mas vale documenta las funciones.

MatiasScantamburlo committed:
fix: separamos test_list_games de test_sanity

MatiasScantamburlo committed:
fix: separamos test_list_games de test_sanity

MatiasScantamburlo committed:
fix: comentamos el test_sanity

- Como ya no hay ednpoint para este test, simplemte lo comentamos

 Cambios a ser confirmados:
	modificados:     tests/test_sanity.py

MatiasScantamburlo committed:
fix: comentamos el test_sanity

- Como ya no hay ednpoint para este test, simplemte lo comentamos

 Cambios a ser confirmados:
	modificados:     tests/test_sanity.py

MatiasScantamburlo committed:
fix: tomamos las sugerencia del PR

MatiasScantamburlo committed:
fix: tomamos las sugerencia del PR

KaiserOfNone commented:
Ahora ya esta pero el formato para los commits no es `feature` o `fix`, es `<ticket-id>`.

KaiserOfNone commented on tests/test_sanity.py:
```diff
@@ -7,7 +7,7 @@
-    response = client.get("/api/borrame")
-    asserts.assert_equal(response.status_code, 200)
-    asserts.assert_equal(response.json(), {"games": []})
+# def test_borrame():
+#     response = client.get("/api/borrame")
+#     asserts.assert_equal(response.status_code, 200)
+#     asserts.assert_equal(response.json(), {"games": []})
```
Siendo que esto era para hacer de ejemplo lo podes borrar del todo.

MatiasScantamburlo committed:
test: Se agrega los test para get_available

MatiasScantamburlo committed:
test: Se agrega los test para get_available

MatiasScantamburlo committed:
Merge branch 'feature_SWC-13_Crear_endpoint_para_enviar_partidas_disponibles' of github.com:HoodedHackers/Backend into feature_SWC-13_Crear_endpoint_para_enviar_partidas_disponibles
